src/censored.jl

```diff
@@ -0,0 +1,168 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+
+function censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+    l <= u || error("the lower bound must be less or equal than the upper bound")
+    Censored(d, l, u)
+end
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
```

Comment:
Is there a need to allow a Float65 lower bound (e.g. `-Inf`) and an `Int` upper bound (or vice versa)?

Reply:
I don't think so; Distributions almost everywhere coerces the parameters to be the same. A does `Truncated` which is very similar to this.

Reply:
So for truncating a discrete distribution on only one side we use `typemin(Int)` or `typemax(Int)`?

Reply:
No, we use eg. 3.0 (or 2.7 etc.) and Inf internally.

Reply:
Users can however create them with `truncated(Poisson(), 3, Inf)`